### PR TITLE
Regression raising Error when having dup cols with single dtype for read csv

### LINKED ISSUE #42022: BUG: read_csv now errors with non-dict dtype and same-name columns
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```Python
with open("test.csv", "w") as cf:
    cf.write("A,B,B\n1,1,1")

pandas.read_csv("test.csv", engine="python", dtype=str)  # used to work, no longer works
pandas.read_csv("test.csv", engine="python", dtype="str")  # used to work, no longer works
```

Now produces 
```
  File ".../pandas/io/parsers/python_parser.py", line 427, in _infer_columns
    and self.dtype.get(old_col) is not None
AttributeError: 'str' object has no attribute 'get'
```

#### Problem description

Documentation for read_csv says that dtype may be a "type name or dict", but the new parser behavior demands a dict in this scenario.

I believe the bug was introduced in https://github.com/pandas-dev/pandas/commit/76792f10a32d157d47721731c5ce2a3192af73f2

#### Expected Output
A dataframe using str dtypes

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2dd9e9ba9009a40191c0c0b96262fa3939d609f0
python           : 3.9.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Mon Apr 12 20:57:45 PDT 2021; root:xnu-6153.141.28.1~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0rc1
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 56.2.0
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 3.0.1
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.3.20
tables           : None
tabulate         : 0.8.7
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #42022
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

regression on master, so no whatsnew

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -689,6 +689,7 @@ cdef class TextReader:
                                 count = counts.get(name, 0)
                             if (
                                 self.dtype is not None
+                                and isinstance(self.dtype, dict)
```
Comment (jreback) [inline pandas/_libs/parsers.pyx:692]: can we use is_dict_like?

**Review by jreback — CHANGES_REQUESTED** [review]
is this for 1.3?

**Comment by jreback** [comment]
> is this for 1.3?

looks like it

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/parsers.pyx]: Yes, thx

**Comment by phofl** [comment]
Yes this is a regression on the rc

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**